src/Core/Services/CryptoService.cs

```diff
@@ -36,189 +34,268 @@ public class CryptoService : ICryptoService
             {
-                return Task.FromResult(_encKey);
+                decUserKey = await DecryptToBytesAsync(encUserKey, masterKey);
             }
-            if (_getEncKeysTask != null && !_getEncKeysTask.IsCompleted && !_getEncKeysTask.IsFaulted)
+            else if (encUserKey.EncryptionType == EncryptionType.AesCbc256_HmacSha256_B64)
             {
-                return _getEncKeysTask;
+                var newKey = await StretchKeyAsync(masterKey, keyBytes => new MasterKey(keyBytes));
+                decUserKey = await DecryptToBytesAsync(encUserKey, newKey);
             }
-            async Task<SymmetricCryptoKey> doTask()
+            else
             {
-                try
-                {
-                    var encKey = await _stateService.GetEncKeyEncryptedAsync();
-                    if (encKey == null)
-                    {
-                        return null;
-                    }
+                throw new Exception("Unsupported encKey type.");
+            }
 
-                    if (key == null)
-                    {
-                        key = await GetKeyAsync();
-                    }
-                    if (key == null)
-                    {
-                        return null;
-                    }
+            if (decUserKey == null)
+            {
+                return null;
+            }
+            return new UserKey(decUserKey);
+        }
 
-                    byte[] decEncKey = null;
-                    var encKeyCipher = new EncString(encKey);
-                    if (encKeyCipher.EncryptionType == EncryptionType.AesCbc256_B64)
-                    {
-                        decEncKey = await DecryptToBytesAsync(encKeyCipher, key);
-                    }
-                    else if (encKeyCipher.EncryptionType == EncryptionType.AesCbc256_HmacSha256_B64)
-                    {
-                        var newKey = await StretchKeyAsync(key);
-                        decEncKey = await DecryptToBytesAsync(encKeyCipher, newKey);
-                    }
-                    else
-                    {
-                        throw new Exception("Unsupported encKey type.");
-                    }
+        public async Task<Tuple<SymmetricCryptoKey, EncString>> MakeDataEncKeyAsync(UserKey key)
+        {
+            if (key == null)
+            {
+                throw new Exception("No user key provided");
+            }
 
-                    if (decEncKey == null)
-                    {
-                        return null;
-                    }
-                    _encKey = new SymmetricCryptoKey(decEncKey);
-                    return _encKey;
-                }
-                finally
-                {
-                    _getEncKeysTask = null;
-                }
+            var newSymKey = await _cryptoFunctionService.RandomBytesAsync(64);
+            return await BuildProtectedSymmetricKey(key, newSymKey, keyBytes => new SymmetricCryptoKey(keyBytes));
+        }
```

Comment:
I was trying to get some type safety for the key when calling these methods, but you're right in that its a lot of code duplication. I went ahead and combined these using a generic in [b1eb263](https://github.com/bitwarden/mobile/pull/2633/commits/b1eb263fef39e401b3608202562fb60e80d4b921). I used a generic instead of `SymmetricCryptoKey` as I'd like to keep the type information here so that others can't accidentally create data encryption keys using other keys.

Reply:
You can keep the type as well using `SymmetricCryptoKey`, giving that you use inheritance for the other keys you can just do:
```c#
public async Task<Tuple<SymmetricCryptoKey, EncString>> MakeDataEncKeyAsync(SymmetricCryptoKey key)
{
    if (key is null)
    {
        throw new ArgumentNullException(nameof(key));
    }

    if (!(key is UserKey) && !(key is OrgKey))
    {
        throw new ArgumentException($"Data encryption keys must be of type UserKey or OrgKey. {key.GetType().FullName} unsupported.");
    }
    // or you can use
    if (!(key is UserKey || key is OrgKey))
    {
        throw new ArgumentException($"Data encryption keys must be of type UserKey or OrgKey. {key.GetType().FullName} unsupported.");
    }
    //...
}

```

Reply:
I've been working in js for too long: [1e8ed1b](https://github.com/bitwarden/mobile/pull/2633/commits/1e8ed1b5ce8b9a67800d4130d7993eeafc7146e6)